keras_cv/layers/preprocessing/mosaic.py

```diff
@@ -72,159 +86,98 @@ def __init__(
-            lambda index: self._update_image(
-                images, permutation_order, mosaic_centers, index
-            ),
-            tf.range(batch_size),
-        )
-
-        if labels is not None:
-            labels = tf.vectorized_map(
-                lambda index: self._update_label(
-                    images, labels, permutation_order, mosaic_centers, index
-                ),
-                tf.range(batch_size),
-            )
-            inputs["labels"] = labels
-
-        if bounding_boxes is not None:
-            # values to translate the boxes by in the mosaic image
-            translate_x = tf.stack(
-                [
-                    mosaic_centers_x - input_width,
-                    mosaic_centers_x,
-                    mosaic_centers_x - input_width,
-                    mosaic_centers_x,
-                ],
-                axis=-1,
-            )
-
-            translate_y = tf.stack(
-                [
-                    mosaic_centers_y - input_height,
-                    mosaic_centers_y - input_height,
-                    mosaic_centers_y,
-                    mosaic_centers_y,
-                ],
-                axis=-1,
-            )
-
-            bounding_boxes = bounding_box.to_dense(bounding_boxes)
-            bounding_boxes = tf.map_fn(
-                lambda index: self._update_bounding_box(
-                    images,
-                    bounding_boxes,
-                    permutation_order,
-                    translate_x,
-                    translate_y,
-                    index,
-                ),
-                tf.range(batch_size),
-                fn_output_signature={
-                    "boxes": tf.RaggedTensorSpec(
-                        shape=[None, 4],
-                        ragged_rank=1,
-                        dtype=self.compute_dtype,
-                    ),
-                    "classes": tf.RaggedTensorSpec(
-                        shape=[None], dtype=self.compute_dtype
-                    ),
-                },
-            )
-            bounding_boxes = bounding_box.to_ragged(bounding_boxes)
-            inputs["bounding_boxes"] = bounding_boxes
-        inputs["images"] = images
-        return inputs
+        return {
+            "permutation_order": permutation_order,
+            "mosaic_centers": mosaic_centers,
+        }
 
-    def _augment(self, inputs):
+    def augment_ragged_image(self, image, transformation, **kwargs):
         raise ValueError(
-            "Mosaic received a single image to `call`.  The layer relies on "
-            "combining multiple examples, and as such will not behave as "
-            "expected.  Please call the layer with 4 or more samples."
+            "Mosaic received ragged images to `call`. The layer relies on "
+            "combining multiple examples with same size, and as such will not "
+            "behave as expected. Please call the layer with dense images."
```

Comment:
maybe add a note that if you would find this method useful open an issue on KerasCV.

Its actually possible to support this - just an engineering investment :)

Reply:
(basically lets make it clear that this is an implementation constraint - not an algorithm constraint)

Reply:
I will open another PR to add this note!